commit 2abeb1802e3c005b17a7123e382171b3fb665971
Author: Gert Wollny <gert.wollny@collabora.com>
Date:   Tue Oct 8 17:27:01 2019 +0200

    vrend: check that the transfer iov holds enough data for the data upload
    
    Closes #140
    
    Signed-off-by: Gert Wollny <gert.wollny@collabora.com>
    Reviewed-by: Emil Velikov <emil.velikov@collabora.com>

Index: virglrenderer-virglrenderer-0.8.0/src/vrend_renderer.c
===================================================================
--- virglrenderer-virglrenderer-0.8.0.orig/src/vrend_renderer.c
+++ virglrenderer-virglrenderer-0.8.0/src/vrend_renderer.c
@@ -6877,15 +6877,22 @@ static int vrend_renderer_transfer_write
             invert = true;
       }
 
+      send_size = util_format_get_nblocks(res->base.format, info->box->width,
+                                          info->box->height) * elsize;
+      if (res->target == GL_TEXTURE_3D ||
+          res->target == GL_TEXTURE_2D_ARRAY ||
+          res->target == GL_TEXTURE_CUBE_MAP_ARRAY)
+          send_size *= info->box->depth;
+
       if (need_temp) {
-         send_size = util_format_get_nblocks(res->base.format, info->box->width,
-                                             info->box->height) * elsize * info->box->depth;
          data = malloc(send_size);
          if (!data)
             return ENOMEM;
          read_transfer_data(iov, num_iovs, data, res->base.format, info->offset,
                             stride, layer_stride, info->box, invert);
       } else {
+         if (send_size > iov[0].iov_len - info->offset)
+            return EINVAL;
          data = (char*)iov[0].iov_base + info->offset;
       }
 
